authentication/php/phpleague/src/PhpLeagueAccessTokenProvider.php

```diff
@@ -0,0 +1,115 @@
+<?php
+/**
+ * Copyright (c) Microsoft Corporation.  All Rights Reserved.
+ * Licensed under the MIT License.  See License in the project root
+ * for license information.
+ */
+
+
+namespace Microsoft\Kiota\Authentication;
+
+
+use Http\Promise\FulfilledPromise;
+use Http\Promise\Promise;
+use Http\Promise\RejectedPromise;
+use League\OAuth2\Client\Provider\GenericProvider;
+use League\OAuth2\Client\Token\AccessToken;
+use Microsoft\Kiota\Abstractions\Authentication\AccessTokenProvider;
+use Microsoft\Kiota\Abstractions\Authentication\AllowedHostsValidator;
+
+/**
+ * Class PhpLeagueAccessTokenProvider
+ * @package Microsoft\Kiota\Authentication
+ * @copyright 2022 Microsoft Corporation
+ * @license https://opensource.org/licenses/MIT MIT License
+ * @link https://developer.microsoft.com/graph
+ */
+class PhpLeagueAccessTokenProvider implements AccessTokenProvider
+{
+    /**
+     * @var TokenRequestContext {@link TokenRequestContext}
+     */
+    private TokenRequestContext $tokenRequestContext;
+    /**
+     * @var AllowedHostsValidator Validates whether a token should be fetched for a request url
+     */
+    private AllowedHostsValidator $allowedHostsValidator;
+    /**
+     * @var array<string, string>
+     */
+    private array $scopes;
+    /**
+     * @var AccessToken|null Token object to re-use before expiry
+     */
+    private ?AccessToken $cachedToken = null;
+    /**
+     * @var GenericProvider OAuth 2.0 provider from PHP League library
+     */
+    private GenericProvider $oauthProvider;
+
+    /**
+     * Creates a new instance
+     * @param TokenRequestContext $tokenRequestContext
+     * @param array $scopes
+     * @param array $allowedHosts
+     */
+    public function __construct(TokenRequestContext $tokenRequestContext, array $scopes, array $allowedHosts = [])
+    {
+
+        $this->tokenRequestContext = $tokenRequestContext;
+        if (empty($scopes)) {
+            throw new \InvalidArgumentException("Scopes cannot be empty");
+        }
+        $this->scopes = $scopes;
+        $this->allowedHostsValidator = new AllowedHostsValidator($allowedHosts);
+        $this->oauthProvider = new GenericProvider([
+            'urlAccessToken' => "https://login.microsoftonline.com/{$tokenRequestContext->getTenantId()}/oauth2/v2.0/token",
+            'urlAuthorize' => '', // required to instantiate Generic Provider
```

Comment:
We wouldn't need this since we're not making a request to get the authorization code in this package. Our interest is in the token endpoint.

Reply:
We could however consider extending this GenericProvider with all necessary endpoints etc so that customers can use our provider to fetch the auth code.

Reply:
I think asking for the auth code from the app developer and getting the token from it is only going to get them halfway to using those flows.
Now, if we think about the different scenarios:
- web application: it's going to use a combination of redirections and they'll have to provide the auth code from the web context (the request URL after the redirection). How can we make it simpler? or document it properly? (maybe this is where a package like identity.web to bridge the gaps between the auth lib and the web framework comes in?)
- client applications: we could have a browser token context, that'd open the browser, go through the steps and capture the auth code automatically.

Reply:
(Web applications)
If we choose to get the auth code for the app developer, at what point would our library would make the auth code redirect?... is it during initialisation of the AuthCodeContext? 
My initial thinking was that a dev could do the redirect themselves with a similar approach to [this PHP League Auth code sample](https://oauth2-client.thephpleague.com/usage/) then they'd initialise the auth provider with the code. We could expose a static method on AuthCodeContext to return the auth code url given a tenantId

(Client applications)
What kind of PHP client applications would require to open a browser? 
Can only think of CLI tools/scripts which could perhaps use a device code flow? 
PHP isn't used for desktop/mobile applications.

Reply:
(Web applications)
If we choose to get the auth code for the app developer, at what point would our library make the auth code redirect?... in the accessTokenProvider?

My initial thinking was that a dev could do the redirect themselves with a similar approach to [this PHP League Auth code sample](https://oauth2-client.thephpleague.com/usage/) then they'd initialise the auth provider with the code. 
Then we can figure out how to expose the PHP League provider we initialise under the hood

Also MSAL's AuthorizationCodeCredential expects a code to be provided